spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
+                                                     untrustedHeight,
+                                                     trustThreshold,
+                                                     trustingPeriod,
+                                                     clockDrift,
+                                                     now)
 
-### Functions
+    if err != nil return (err, trustedState)
 
-The function _Bisection_ checks whether to trust header _h2_ based on the trusted header _h1_. It does so by calling
-the function _CheckSupport_ in the process of
-bisection/recursion. _CheckSupport_ implements the trusted period method and, for two adjacent headers (in term of heights), it checks uninterrupted sequence of proof.
+    now = System.Time()
+    if !isWithinTrustedPeriod(trustedHeader, trustingPeriod, now) {
+        return (ErrHeaderNotWithinTrustedPeriod, trustedState)
+    }
 
-_Assumption_: In the following, we assume that _h2.Header.height > h1.Header.height_. We will quickly discuss the other case in the next section.
+    return (nil, newTrustedState)
+}
+```
 
-We consider the following set-up:
+Note that in case `VerifyAndUpdateSingle` returns without an error (untrusted header
+is successfully verified) then we have a guarantee that the transition of the trust
+from `trustedState` to `newTrustedState` happened during the trusted period of
+`trustedState.SignedHeader.Header`.
 
-- the lite client communicates with one full node
-- the lite client locally stores all the signed headers it obtained (trusted or not). In the pseudo code below we write _Store(header)_ for this.
-- If _Bisection_ returns _false_, then the lite client has seen a forged header.
+**VerifyAndUpdateBisection.** The function `VerifyAndUpdateBisection` implements
+recursive logic for checking if it is possible building trust
+relationship between trustedState and untrusted header at the given height over
+finite set of (downloaded and verified) headers.
 
-  - However, it does not know which header(s) is/are the problematic one(s).
-  - In this case, the lite client can submit (some of) the headers it has seen as evidence. As the lite client communicates with one full node only when executing Bisection, there are two cases
-    - the full node is faulty
-    - the full node is correct and there was a fork in Tendermint consensus. Header _h1_ is from a different branch than the one taken by the full node. This case is not focus of this document, but will be treated in the document on fork accountability.
+```go
+func VerifyAndUpdateBisection(trustedState TrustedState,
+                              untrustedHeight int64,
+                              trustThreshold TrustThreshold,
+                              trustingPeriod Duration,
+                              clockDrift Duration,
+                              now Time) (error, TrustedState) {
+
+    trustedHeader = trustedState.SignedHeader.Header
+    assert trustedHeader.Height < untrustedHeight
```

Comment:
Aren't this and the one below redundant with checks that happen in verifySingle?

Reply:
Good point.